src/buffer.rs

```diff
@@ -0,0 +1,294 @@
+    /// ```
+    pub fn get(&self) -> &[u8] {
+        &self.buffer[..=self.write_end]
+    }
+}
+
+impl Index<usize> for DynamicBuffer {
+    type Output = u8;
+
+    fn index(&self, index: usize) -> &Self::Output {
+        self.buffer.index(self.start_offset + index)
+    }
+}
+impl IndexMut<usize> for DynamicBuffer {
+    fn index_mut(&mut self, mut index: usize) -> &mut Self::Output {
+        index += self.start_offset;
+
+        if let Some(diff) = index.checked_sub(self.buffer.len()) {
+            self.buffer
+                .extend(core::iter::repeat(0).take(core::cmp::max(self.alloc_increment, diff + 1)));
+        }
+
+        self.write_end = self.write_end.max(index);
+
+        self.buffer.index_mut(index)
+    }
+}
+
+impl<Ctx: Copy, E> Pwrite<Ctx, E> for DynamicBuffer {
+    type Buffer = Self;
+
+    fn pwrite_with<N: TryIntoCtx<Ctx, Self::Buffer, Error = E>>(
+        &mut self,
+        n: N,
+        offset: usize,
+        ctx: Ctx,
+    ) -> Result<usize, E> {
+        self.start_offset += offset;
+
+        let end = n.try_into_ctx(self, ctx)?;
+
+        self.start_offset -= offset;
+
+        Ok(end)
+    }
+}
+
+macro_rules! num_impl {
+    ($t:ty) => {
+        impl TryIntoCtx<crate::Endian, DynamicBuffer> for $t {
+            type Error = crate::Error; // doesn't matter, this is infallible
+
+            fn try_into_ctx(
+                self,
+                buf: &mut DynamicBuffer,
+                ctx: Endian,
+            ) -> Result<usize, Self::Error> {
+                for (index, byte) in IntoIterator::into_iter(if ctx.is_little() {
+                    self.to_le_bytes()
+                } else {
+                    self.to_be_bytes()
+                })
+                .enumerate()
+                {
+                    buf[index] = byte;
+                }
+
+                Ok(core::mem::size_of::<Self>())
+            }
+        }
+
+        impl TryIntoCtx<crate::Endian, DynamicBuffer> for &$t {
+            type Error = crate::Error;
+
+            fn try_into_ctx(
+                self,
+                buf: &mut DynamicBuffer,
+                ctx: Endian,
+            ) -> Result<usize, Self::Error> {
+                (*self).try_into_ctx(buf, ctx)
+            }
+        }
+    };
+}
+
+num_impl!(i8);
+num_impl!(i16);
+num_impl!(i32);
+num_impl!(i64);
+num_impl!(i128);
+
+num_impl!(u8);
+num_impl!(u16);
+num_impl!(u32);
+num_impl!(u64);
+num_impl!(u128);
+
+num_impl!(f32);
+num_impl!(f64);
+
+#[cfg(test)]
+mod tests {
+    use super::DynamicBuffer;
+    use crate::ctx::TryIntoCtx;
+    use crate::{Endian, Pwrite};
+
+    struct Test {
+        a: u16,
+        b: u32,
+        c: u64,
+    }
+
+    #[test]
+    fn int_write() {
+        let mut buf = DynamicBuffer::new();
+
+        buf.pwrite_with(0x1234u16, 0, Endian::Little).unwrap();
+        buf.pwrite_with(0x5678i16, 2, Endian::Big).unwrap();
+
+        assert_eq!(buf.get(), [0x34, 0x12, 0x56, 0x78]);
```

Comment:
i haven't thought deeply, but be careful to think about how this test would pass on big endian machine. In other places in scroll there sometimes need to be `cfg(target_endian = "little")` etc.; I think you're ok here but something to think about for a minute, because in my experience easy to make subtle mistake with endianness in tests and it fails years later when someone tries to build it on some big endian machine :)